streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskSchedulerImpl.scala

```diff
@@ -0,0 +1,178 @@
+import org.apache.gearpump.streaming.appmaster.TaskSchedulerImpl.TaskLaunchData
+import org.apache.gearpump.streaming.task.TaskId
+import org.apache.gearpump.util.LogUtil
+import org.slf4j.Logger
+
+import scala.collection.mutable
+
+/**
+ * This schedules tasks to run for new allocated resources.
+ */
+trait TaskScheduler {
+
+  /**
+   * This notify the scheduler that the task DAG is created.
+   * @param dag task dag
+   */
+  def setTaskDAG(dag: DAG): Unit
+
+  /**
+   * Get the resource requirements for all unscheduled tasks.
+   */
+  def getResourceRequests(): Array[ResourceRequest]
+
+  /**
+   * This notify the scheduler that a resource slot on {workerId} and {executorId} is allocated, and
+   * expect a task to run in return.
+   * Task locality should be considered when deciding whether to offer a task on target {worker}
+   * and {executor}
+   *
+   * @param workerId which worker this resource is located.
+   * @param executorId which executorId this resource belongs to.
+   * @return
+   */
+  def resourceAllocated(workerId : Int, executorId: Int) : Option[TaskLaunchData]
+
+  /**
+   * This notify the scheduler that {executorId} is failed, and expect a set of
+   * ResourceRequest for all failed tasks on that executor.
+   *
+   * @param executorId executor that failed
+   * @return
+   */
+  def executorFailed(executorId: Int) : Array[ResourceRequest]
+}
+
+class TaskSchedulerImpl(appId : Int, config: Config)  extends TaskScheduler {
+  private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+
+  private var tasks = Map.empty[TaskId, TaskLaunchData]
+
+  // maintain a list to tasks to be scheduled
+  private var tasksToBeScheduled = Map.empty[Locality, mutable.Queue[TaskLaunchData]]
+
+  // find the locality of the tasks
+  private val taskLocator = new TaskLocator(config)
+
+  private var executorToTaskIds = Map.empty[Int, Set[TaskId]]
+
+  def setTaskDAG(dag: DAG): Unit = {
+    dag.tasks.foreach { params =>
+      val (taskGroupId, taskDescription) = params
+      0.until(taskDescription.parallelism).map((taskIndex: Int) => {
+        val taskId = TaskId(taskGroupId, taskIndex)
+        val locality = taskLocator.locateTask(taskDescription)
+        val taskLaunchData = TaskLaunchData(taskId, taskDescription, dag.subGraph(taskGroupId))
+        tasks += (taskId -> taskLaunchData)
+        scheduleTaskLater(taskLaunchData, locality)
+      })
+    }
+    val nonLocalityQueue = tasksToBeScheduled.getOrElse(NonLocality, mutable.Queue.empty[TaskLaunchData])
+    //reorder the tasks to distributing fairly on workers
+    val taskQueue = nonLocalityQueue.sortBy(_.taskId.index)
+    tasksToBeScheduled += (NonLocality -> taskQueue)
+  }
+
+  def getResourceRequests(): Array[ResourceRequest] ={
+    fetchResourceRequests(fromOneWorker = false)
+  }
+
+  import Relaxation._
+  private def fetchResourceRequests(fromOneWorker: Boolean = false): Array[ResourceRequest] ={
```

Comment:
The task scheduler to far away to be optimized, and I expect the TaskSchedule interface will still need large change. 

There are several  requirements to be met:
1. Need to response to DAG change.
2. Need to response to task fail, task create.
3. Need to response to executor fail.
4. Need to response to user policy change.

So, ideally, 
The Taskscheduler need to reponse to four event source.
DAG, task, executor, user.

And when one event from above event source happen, the task scheduler should request new resource or schedule new tasks accordingly. 

Howerver, the objective of this PR is to seperate the responsibility and remove the coupling in AppMaster. The AppMaster, TaskManager, and ExecutorManager are new written, the TaskScheduler are copied from old TaskSet impl with interface improvement. 

I will try to leave exising code implementation unchanged if possible.